test/pkg.jl

```diff
@@ -398,4 +398,73 @@ temp_pkg_dir() do
         @test contains(msg, "$package is not an installed package")
         @test !contains(msg, "signal (15)")
     end
+
+    # partial Pkg.update
+    begin
+        nothingtodomsg = "INFO: No packages to install, update or remove\n"
+
+        ret, out, err = @grab_outputs begin
+            Pkg.rm("Example")
+            Pkg.add("Example")
+        end
+        @test ret == nothing && out == ""
+        @test contains(err, "INFO: Installing Example v")
+
+        ret, out, err = @grab_outputs Pkg.update("Example")
+        @test ret == nothing && out == ""
+        @test contains(err, nothingtodomsg)
+
+        ret, out, err = @grab_outputs begin
+            Pkg.rm("Example")
+            Pkg.add("Example", v"0", v"0.4.1-")
+        end
+        @test ret == nothing && out == ""
+        @test contains(err, "INFO: Installing Example v0.4.0")
+
+        ret, out, err = @grab_outputs Pkg.update("Example")
+        @test ret == nothing && out == ""
+        @test ismatch(r"INFO: Package Example was set to version 0\.4\.0, but a higher version \d+\.\d+\.\d+\S* exists.\n", err)
+        @test contains(err, "The update is prevented by explicit requirements constraints. Edit your REQUIRE file to change this.\n")
+        @test contains(err, nothingtodomsg)
+
+        ret, out, err = @grab_outputs begin
+            Pkg.rm("Example")
+            Pkg.add("Example")
+            Pkg.pin("Example", v"0.4.0")
+        end
+        @test ret == nothing && out == ""
+        @test contains(err, "INFO: Installing Example")
+
+        ret, out, err = @grab_outputs Pkg.update("Example")
+        @test ret == nothing && out == ""
+        @test contains(err, "INFO: Package Example: skipping update (pinned)...\n")
+        @test ismatch(r"INFO: Package Example was set to version 0\.4\.0, but a higher version \d+\.\d+\.\d+\S* exists.\n", err)
+        @test contains(err, "The package is fixed. You can try using `Pkg.free(\"Example\")` to update it.")
+        @test contains(err, nothingtodomsg)
+
+        metadata_dir = joinpath(Pkg.Dir.path(), "METADATA")
+        const test_commit = "313bfaafa301e82d40574a778720e893c559a7e2"
+
+        LibGit2.with(LibGit2. GitRepo, metadata_dir) do repo
+            LibGit2.reset!(repo, LibGit2.Oid(test_commit), LibGit2.Consts.RESET_HARD)
```

Comment:
The following calls to `Pkg.update` take care of that: even if the package update is partial, METADATA is always updated to the latest version.